crates/ruff_server/src/edit/notebook.rs

```diff
@@ -0,0 +1,189 @@
+use std::{collections::HashMap, hash::BuildHasherDefault};
+
+use anyhow::Ok;
+use lsp_types::{NotebookCellKind, Url};
+use rustc_hash::FxHashMap;
+
+use crate::{PositionEncoding, TextDocument};
+
+use super::DocumentVersion;
+
+/// The state of a notebook document in the server. Contains an array of cells whose
+/// contents are internally represented by [`TextDocument`]s.
+#[derive(Clone, Debug)]
+pub(crate) struct NotebookDocument {
+    cells: Vec<NotebookCell>,
+    metadata: ruff_notebook::RawNotebookMetadata,
+    version: DocumentVersion,
+    // Used to quickly find the index of a cell for a given URL.
+    cell_index: FxHashMap<lsp_types::Url, usize>,
+}
+
+#[derive(Clone, Debug)]
+struct NotebookCell {
```

Comment:
I think it will be useful to add some documentation on the struct, fields and it's various methods. There are lots of methods and it's useful as a developer to know in what scenario would a certain method be useful 😅